cEP-0028.md

```diff
@@ -0,0 +1,428 @@
+
+## Proposal
+
+The optimization process focuses to maintain minimal changes to coala
+core and rather optimize the coala and language server interface. The
+following are the proposed ideas.
+
+1. Section Tagging<br/>
+   Some sections are not often required to be executed on every cycle of
+   coala analysis especially during use cases such as this. The parameter
+   set can thus be extended to support a parameter named `tags`. Since
+   these parameters are not associated with any particular bear but rather
+   control the effect of various sections during executing coala, no support
+   extension for bears would be necessary. These parameters enable support
+   for hooking a section for a particular tag type. For example:
+
+   ```
+   [longrunning]
+   tags = save
+   enabled = True
+   bears = SomeLongRunningBear
+   ```
+
+   This kind of a configuration will enable this section during analysis
+   cycles of coala via a language server when a save request is raised
+   while not impacting any untagged regular invocations.
+
+   ```
+   [longrunning]
+   files = *.py
+   bears = SomeLongRunningBear
+
+   [longrunning.some]
+   tags += save, change
+   bears = SomeLongRunningBear
+   ```
+
+   This example shows a case where the sections inherit the tags and
+   thereby `longrunning.some` runs on both `save` and `change` tagged
+   execution cycles.
+
+   To enable tagged executional support, coala-core needs to be extended
+   and at least support a new optional INPUT cli argument taking list of
+   tags to be executed. The execution of the sections will be based upon
+   the sections being tagged as at least one of the tags being invoked.
+   This extension of support for `tags` would require at least the
+   following changes:
+
+   - `coalib.settings.Section`<br/>
+     Adds a new method to the Section class to check if
+     the section should be enabled for a particular tag
+     based invocation.
+
+   ```python
+   class Section:
+
+     def is_enabled_for_tags(self, enabled_tags):
+       """
+       Checks if this section is enabled when the execution
+       limits are specified by tags.
+
+       :param enabled_tags: List of all enabled tags.
+       :return:             True or False
+       """
+       if len(enabled_tags) == 0:
+         return True
+
+       set_enabled = set(enabled_tags)
+       set_tags = set(self.get('tags', []))
+
+       matching = set_tags.intersection(set_enabled)
+       return len(matching) > 0
+   ```
+
+   - `coalib.parsing.DefaultArgParser.default_arg_parser`<br/>
+     Add a new `tags` cli argument type to the argparser.
+
+     ```python
+     misc_group.add_argument(
+         '-t', '--tags', nargs='*', type=str,
+         help='tags of sections to execute')
+     ```
+
+   - `coalib.coala_main.run_coala`<br/>
+     Patch `run_coala()` to check if the corresponding tags
+     of the section are enabled.
+
+   ### Supported Tags
+
+   The following are the supported default tags groups:
```

Comment:
This needs to clarify that it is the set of tags which coala-ls will use.

Tags should be able to be used by other components in other ways, as long as they dont conflict with these uses by coala-ls.